tiledb/cloud/dag/dag.py

```diff
@@ -689,44 +832,90 @@ def report_node_complete(self, node: Node):
         the graph information on the server can know about these failed nodes
         even though there will be no ArrayTask for them in our database.
         """
-
-        if node.status == st.Status.COMPLETED:
-            self.completed_nodes[node.id] = node
-            for child in node.children.values():
-                self._exec_node(child)
-            if node.local_mode:
-                to_report = models.ArrayTaskStatus.COMPLETED
-        else:
-            if node.status == st.Status.CANCELLED:
+        with self._lifecycle_condition:
+            # A node may be either "running" or "not started" depending upon
+            # whether it failed or was cancelled.
+            self.running_nodes.pop(node.id, None)
+            self.not_started_nodes.pop(node.id, None)
+
+            if node.status is st.Status.COMPLETED:
+                self.completed_nodes[node.id] = node
+                for child in node.children.values():
+                    self._maybe_exec(child)
```

Comment:
This also happens at line 859; is this deliberate?

Reply:
It is not. It didn't *hurt* anything (`_maybe_exec` is safe to call repeatedly) but definitely not the intended result.